README.md
# cfc

Short for 'Concurrent File Crawler' - A crawler that searches a file system for files matching a specified bash pattern.

## Table of Contents

* [General Info](#general.info)
* [Setup](#setup)
* [Usage](#usage)
  * [About Bash Patterns](#about.bash.patterns)
  * [Command Line Arguments](#command.line.arguments)
* [Contributing](#contributing)
* [License](#license)

<a name="general.info"></a>
## General Info

In today's world, computer storage is becoming larger and cheaper. For laptops, storage can vary from several gigabytes to a couple terabytes, and that range can extends much larger for PCs, up to several terabytes of storage. This can make searching and managing older files and directories very difficult. In Linux, there is the ```find``` command that can assist with this, but the program itself is single-threaded and the syntax is cumbersome.

This application nicknamed ```cfc``` is a more simplified version of the ```find``` command. It will crawl through a file system and search for files & directories that match a given bash pattern specified by the user. It is designed to be easier to use and high-performance through PThreads.

**Examples**

Searching for Makefiles:

![doc/usage_1.png](https://github.com/cvikupitz/cfc/blob/master/doc/usage_1.png)

Counting C source and header files:

![doc/usage_2.png](https://github.com/cvikupitz/cfc/blob/master/doc/usage_2.png)

<a name="setup"></a>
## Setup

* Clone the repository: 
  * ```https://github.com/cvikupitz/cfc.git```
* Build the executable via the Makefile:
  * ```make clean cfc```
* Once built, you can run the executable as such:
  * ```./cfc '*.(c|h)' -I/home -X2 -a```
* Other options can be viewed with the help command:
  * ```make help```

<a name="usage"></a>
## Usage

The file crawler program requires one argument which is the bash pattern to be used for matching files and directories. This is also the only argument that is not preceded by dashes like the command line arguments. Since the crawler requires a bash pattern for matching, it is critical that you pass it in quotes in order to prevent bash itself from interpreting the pattern. For example:

```bash
$ ./cfc '*.txt'
```

is the correct way to run the program. However, let's say you have *test.txt* and *example.txt* in your current working directory. If you run the program without the quotes like this: 

```bash
$ ./cfc *.txt
```

then bash will run the program like this:

```bash
$ ./cfc example.txt test.txt
```

in which case the pattern used for matching becomes *test.txt*. So make sure to always enclose the pattern with quotes!

If you are rusty on bash patterns, see the below section for a brief refresher.

<a name="about.bash.patterns"></a>
### About Bash Patterns

'\*'     - Matches all strings, including the empty/null string.

'?'     - Matches a single character.

'[...]' - Matches any of the enclosed characters. A pair of characters separated by a hyphen denotes a range expression; any character that sorts between those two characters (inclusive) is a match. If the first character following the opening bracket ('[') is a carat ('^'), then any characters NOT in the set is a match.

**Examples**:

'**file?.txt**' matches '**file1.txt**', '**file2.txt**', and '**file8.txt**', but not '**fil1.txt**', '**file.txt**', or '**file10.txt**'.

'***.txt**' matches '**test.txt**', '**.t.txt**', and '**.txt**', but not '**test.pdf**', '**test.png**', or '  '.

'**[0-9]+**' matches '**0**', '**001**', and '**99**', but not '**1A**', '**A1**', or '**TEST**'.

'**\[^0-9]+**' matches '**TEST**', '**(A)**', and '**-B**', but not '**A1**', '**100**', or '**TEST__9**'.

<a name="command.line.arguments"></a>
### Command Line Arguments

Below is an overview of all the available command line arguments supported by the crawler program.

| Argument                     | Default   | Description                                                  |
| ---------------------------- | --------- | ------------------------------------------------------------ |
| ```-a, --all```              |           | The crawler will not ignore 'hidden' files and directories, that is, if the entry starts with '.'. If the entry is a file, the crawler will match the file against the pattern and include in the results if it's a match. If the entry is a directory, then the crawler will traverse down into that folder. |
| ```-c, --conflict```         |           | Performs a 'conflicting' search, that is, all files that do not match the specified bash pattern are considered matches, while entries that do match the bash pattern are ignored. |
| ```-F, --check-folders```    |           | Includes folders in the search. In addition to traversing into sub-folders, the bash pattern will also be applied to the folder names and included in the results if found as a match. |
| ```-I<DIR>, --include=DIR``` | "./"      | Include ```DIR``` in the search path. You may specify multiple search paths by giving multiple flags. If no flags are specified, only the current working directory is crawled. |
| ```-i, --ignore-case```      |           | Performs a case-insensitive search. If the pattern specified is '*\*.txt*', then this flag will cause the files *test.txt* and '*test.TXT*' to match. |
| ```--max-depth=N```          | Unbounded | Does not crawl more than N sub-directories from each directory in the search path. If there exists a directory */home/users/foobar/tests* and this path is included in the search path, and max depth specified is 2, then the crawler will stop searching within */home/users/foobar*. If max depth is set to 0, then that means no sub-folders in */home* will be searched. |
| ```--min-depth=N```          | 0         | The crawler will crawl N number of sub-directories before it will start matching files and folders. If there exists a directory */home/users/foobar/tests* and this path is included in the search path, and min depth specified is 2, then the crawler will only start checking entries in */home/users/foobar*. |
| ```-M<N>, --max-results=N``` | Unbounded | Sets the number of maximum results to display. Since the output is in alphabetical order, this means that the first N results in alphabetical order is displayed. |
| ```-q, --quiet```            |           | Does not display any of the matched results, only the total number of matches. |
| ```-r, --reverse```          |           | Reverses the output ordering of the matched results. By default, all paths are output in alphabetical order. This flag will reverse the alphabetical ordering. |
| ```-W, --no-warn```          |           | All warning messages during the crawling phase are muted. Mostly includes messages related to failing to open additional directories. |
| ```-X<N>, --threads=N```     | 1         | Sets the number of threads to run in the file crawling phase. Note that this does not apply to argument parsing or displaying the matched results. |
| ```-?, --help```             |           | Displays a helpful message along with a list of all arguments, then exits. |
| ```--usage```                |           | Displays the full usage message, then exits.                 |
| ```-V, --version```          |           | Displays the current version, then exits.                    |

<a name="contributing"></a>
## Contributing

If there is a bug or a new feature you think would be useful/cool, feel free to contribute to the project by opening your own branch and pull request:

* Fork the project: https://github.com/cvikupitz/cfc
* Create your own branch:
  * ```git checkout -b feature/foobar```
* Commit your changes:
  * ```git commit -am 'Add some foobar'```
* Push to your new branch:
  * ```git push origin feature/foobar```
* Create a new pull request

Kindly name your new features with the branch name 'feature/*name*' where *name* is a brief name of the feature. For fixes, name the branch 'fix/*name*' with *name* being the name/descriptor of the issue.

<a name="license"></a>
## License

[MIT License](https://github.com/cvikupitz/cfc/blob/master/LICENSE)

CHANGE_LOG.md
#### Version 1.0

* Initial release of CFC
  * Successfully crawls and displays matched file and directory names based on a user-provided regular expression.
  * Ability to perform case-sensitive and case-insensitive searches.
  * Provides a multi-threaded implementation with some number of threads specified by the user.
  * Provides ability to print a maximum number of matches or none at all, also specified by user.
  * Provides ability to crawl up to a maximum depth of sub-folders, specified by user.

#### Version 1.1

* Added *-W*, *--min-depth*, & *--max-depth* arguments.
  * *-W, --no-warn*: User can silence all error/warning messages during the crawling phase (e.g. directories fail to open).
  * *--max-depth*: Renamed from the *-D* flag.
  * *--min-depth*: User can now specify a minimum depth of sub-directories to traverse before results will be matched by the pattern.

src/arg_parser.c
/**
 * MIT License
 *
 * Copyright (c) 2021 Cole Vikupitz
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <argp.h>
#include <stdlib.h>
#include <string.h>
#include "arg_parser.h"
#include "file_utils.h"

static ProgArgs *prog_args = NULL;

/* Program version */
const char *argp_program_version = "cfc 1.1";
/* Address to send bugs */
const char *argp_program_bug_address = "https://github.com/cvikupitz/cfc/issues";
/* Documentation for usage */
static char doc[] =
"Recursively scans for files that match the specified pattern 'REGEX'.\n\
The regex specified must be a bash pattern that will be used to match against the files (e.g., 'test??.txt', 't*.txt', '[a-z].txt'). \
The bash pattern must also be passed inside quotes to prevent bash itself from interpreting the expression before passing it as an argument.\n\n\
You can specify which paths on the system to search in with the -I flag. If no paths are specified, the current working directory ('./') will \
only be searched.\n\n\
A quick refresher on bash patterns:\n\
  '*'     - Matches all strings, including the empty/null string.\n\
  '?'     - Matches a single character.\n\
  '[...]' - Matches any of the enclosed characters. A pair of characters\n\
            separated by a hyphen denotes a range expression; any character\n\
            that sorts between those two characters (inclusive) is a match.\n\
            If the first character following the opening bracket ('[') is a\n\
            carat ('^'), then any characters NOT in the set is a match.\n\n\
Examples:\n\
  'file?.txt' matches 'file1.txt', 'file2.txt', and 'file8.txt', but not\n\
      'fil1.txt', 'file.txt', or 'file10.txt'.\n\
  '*.txt' matches 'test.txt', '.t.txt', and '.txt', but not\n\
      'test.pdf', 'test.png', or '  '.\n\
  '[0-9]+' matches '0', '001', and '99', but not\n\
      '1A', 'A1', or 'TEST'.\n\
  '[^0-9]+' matches 'TEST', '(A)', and ' -B', but not\n\
      'A1', '100', or 'TEST__9'.\v\
All results and other informational output will be printed to standard output, and all error messages (e.g., directory fails to open) will be \
printed to standard error.\n\n\
Exit Status:\n\
  0 if successful\n\
  1 for minor issues (e.g., argument(s) parsed were invalid)\n\
  2 for major issues (e.g., dynamic memory allocation failed)\n";

/*
 * Converts a bash pattern to a regex to be compiled. Stores the result into 'dest'.
 *
 * Performs the following conversions:
 *    - Adds a '^' to the start
 *    - Converts a '.' to a '\.'
 *    - Converts a '?' to a '.'
 *    - Converts a '*' to a '.*'
 *    - Adds a '$' to the end
 */
static void convert_to_bash(char *regex, char dest[]) {

    char *curr = regex;
    int i = 0;

    dest[i++] = '^';
    while (*curr != '\0') {
        switch (*curr) {
            case '.':
                dest[i++] = '\\';
                dest[i++] = '.';
                break;
            case '?':
                dest[i++] = '.';
                break;
            case '*':
                dest[i++] = '.';
                dest[i++] = '*';
                break;
            default:
                dest[i++] = *curr;
                break;
        }
        curr++;
    }
    dest[i++] = '$';
    dest[i] = '\0';
}

/*
 * Function used to parse the program arguments. Iterates through the flags and sets the flags &
 * properties in the struct as needed.
 */
static int parse_options(int key, char *arg, struct argp_state *state) {

    char *after;
    int *arg_count = state->input;

    switch (key) {
        case 'a':
            prog_args->progFlags |= (1 << SHOW_ALL);
            break;
        case 'c':
            prog_args->progFlags |= (1 << CONFLICT);
            break;
        case 'F':
            prog_args->progFlags |= (1 << CHECK_FOLDERS);
            break;
        case 'I':
            {
                int i = prog_args->nPaths;
                if (i >= MAX_DIRS)
                    break;
                strcpy(prog_args->searchPaths[i], arg);
                file_path_append(prog_args->searchPaths[i], '/');
                prog_args->nPaths++;
                break;
            }
        case 'i':
            prog_args->progFlags |= (1 << IGNORE_CASE);
            break;
        case 200:
            {
                int temp = strtol(arg, &after, 10);
                if (temp < 0) {
                    argp_failure(state, 1, 0, "invalid max depth: '%s' - must be an int greater than or equal to 0.", arg);
                } else {
                    prog_args->maxDepth = temp;
                }
                break;
            }
        case 201:
            {
                int temp = strtol(arg, &after, 10);
                if (temp < 0) {
                    argp_failure(state, 1, 0, "invalid min depth: '%s' - must be an int greater than or equal to 0.", arg);
                } else {
                    prog_args->minDepth = temp;
                }
                break;
            }
        case 'X':
            {
                int temp = strtol(arg, &after, 10);
                if (temp <= 0) {
                    argp_failure(state, 1, 0, "invalid thread count: '%s' - must be an int greater than 0.", arg);
                } else {
                    prog_args->nThreads = temp;
                }
                break;
            }
        case 'M':
            {
                long temp = strtol(arg, &after, 10);
                if (temp <= 0) {
                    argp_failure(state, 1, 0, "invalid max results: '%s' - must be an int greater than 0.", arg);
                } else {
                    prog_args->maxResults = temp;
                }
                break;
            }
        case 'q':
            prog_args->progFlags |= (1 << QUIET);
            break;
        case 'r':
            prog_args->progFlags |= (1 << REVERSE);
            break;
        case 'W':
            prog_args->progFlags |= (1 << NO_WARN);
            break;
        case ARGP_KEY_ARG:
            {
                char buffer[BUFFER_SIZE];
                (*arg_count)--;
                convert_to_bash(arg, buffer);
                strcpy(prog_args->regex, buffer);
                break;
            }
        case ARGP_KEY_END:
            if ((*arg_count) > 0) {
                argp_failure(state, 1, 0, "Pattern 'REGEX' is undefined, please specify the pattern for matching.");
            }
            break;
    }
    return 0;
}

/* List of program options parsable by argp library */
static struct argp_option options[] = {
    {0, 0, 0, 0, "Search Options", 1},
    {"all", 'a', 0, 0, "Does not ignore entries starting with '.'", 0},
    {"conflict", 'c', 0, 0, "Performs a conflicting search; matches all that do NOT match the specified pattern", 0},
    {"check-folders", 'F', 0, 0, "Includes folders in the search", 0},
    {"include", 'I', "DIR", 0, "Adds DIR to the search path", 0},
    {"ignore-case", 'i', 0, 0, "Performs a case-insensitive search", 0},
    {"max-depth", 200, "N", 0, "Recursively searches no more than N subdirectories for each directory in the search path", 0},
    {"min-depth", 201, "N", 0, "Only search for matches that are at within least N subdirectories for each directory in the search path", 0},
    {"threads", 'X', "N", 0, "Performs the search with N number of PThreads", 0},
    {0, 0, 0, 0, "Output Options", 2},
    {"max-results", 'M', "N", 0, "Display no more than N results", 0},
    {"quiet", 'q', 0, 0, "Prints only the number of matches, not the matches themselves", 0},
    {"reverse", 'r', 0, 0, "Reverses the sorting when displaying the matches", 0},
    {"no-warn", 'W', 0, 0, "Suppresses all error & waring messages during file crawling", 0},
    { 0 }
};

static struct argp argps = {options, parse_options, "'REGEX'", doc, 0, 0, 0};

int prog_args_parse(int argc, char **argv, ProgArgs **progArgs) {

    int result, arg_count = 1;

    if ((prog_args = (ProgArgs *)malloc(sizeof(ProgArgs))) == NULL) {
        fprintf(stderr, "ERROR: Failed to allocate enough emmory from heap.\n");
        return 1;
    } else {
        prog_args->nPaths = 0;
        prog_args->maxDepth = -1;
        prog_args->minDepth = 0;
        prog_args->maxResults = 0;
        prog_args->nThreads = 1;
        prog_args->progFlags = 0;
    }

    if ((result = argp_parse(&argps, argc, argv, 0, 0, &arg_count)) == 0) {
        *progArgs = prog_args;
    } else {
        free(prog_args);
    }

    return result;
}
